sdk/core/azure-amqp/_connection.py

```diff
@@ -831,24 +833,25 @@ def create_session(
         :return: A new Session.
         :rtype: ~pyamqp._session.Session
         """
-        assigned_channel = self._get_next_outgoing_channel()
-        kwargs['offered_capabilities'] = offered_capabilities
-        session = Session(
-            self,
-            assigned_channel,
-            name=name,
-            handle_max=handle_max,
-            properties=properties,
-            next_outgoing_id=next_outgoing_id,
-            incoming_window=incoming_window,
-            outgoing_window=outgoing_window,
-            desired_capabilities=desired_capabilities,
-            allow_pipelined_open=allow_pipelined_open or self._allow_pipelined_open,
-            idle_wait_time=idle_wait_time or self._idle_wait_time,
-            network_trace= network_trace or self._network_trace,
-            network_trace_params=dict(self._network_trace_params),
-            **kwargs,
-        )
+        with self._connection_lock:
```

Comment:
Added a lock only on create session as the outgoing channel in incremented. Didnt feel the need to add it anywhere else as nothing else from the outside is changing connection state or any other resource from outside.

Reply:
What about `open` and `close`? They're both infrequent enough that I think locking wouldn't be a burden, but could also prevent some weird state-change scenarios.

Reply:
One thing we should analyze in general is how inbetween-states are handled in the multi-threaded scenario. e.g. opening isn't an instant process, it involves several state changes and frame interactions - so what happens if another thread interrupts this. Same question for disconnecting, reconnecting and error recovery, or failure/unrecoverable error.